drivers/counter/counter_esp32_rtc.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2022 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_rtc_timer
+
+/*
+ * Include esp-idf headers first to avoid
+ * redefining BIT() macro
+ */
+#include "soc/rtc_cntl_reg.h"
+#include "soc/rtc.h"
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/counter.h>
+
+#if defined(CONFIG_SOC_ESP32C3)
+#include <zephyr/drivers/interrupt_controller/intc_esp32c3.h>
+#else
+#include <zephyr/drivers/interrupt_controller/intc_esp32.h>
+#endif
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(esp32_counter_rtc, CONFIG_COUNTER_LOG_LEVEL);
+
+#if defined(CONFIG_SOC_ESP32C3)
+#define ESP32_COUNTER_RTC_ISR_HANDLER isr_handler_t
+#else
+#define ESP32_COUNTER_RTC_ISR_HANDLER intr_handler_t
+#endif
+
+static void counter_esp32_isr(void *arg);
+
+struct counter_esp32_config {
+	struct counter_config_info counter_info;
+	int irq_source;
+};
+
+struct counter_esp32_data {
+	struct counter_alarm_cfg alarm_cfg;
+	uint32_t ticks;
+};
+
+static int counter_esp32_init(const struct device *dev)
+{
+	const struct counter_esp32_config *cfg = dev->config;
+
+	/* SLOW_CK is the default clk source */
+	if (cfg->counter_info.freq != rtc_clk_slow_freq_get_hz()) {
+		return -EINVAL;
+	}
+
+	esp_intr_alloc(cfg->irq_source,
+			0,
+			(ESP32_COUNTER_RTC_ISR_HANDLER)counter_esp32_isr,
+			(void *)dev,
+			NULL);
+
+	return 0;
+}
+
+static int counter_esp32_start(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* RTC main timer runs after power-on reset */
+	return 0;
+}
+
+static int counter_esp32_stop(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/*
+	 * Any reset/sleep mode, except for the power-up
+	 * reset, will not stop or reset the RTC timer
+	 * ESP32 TRM v4.6 sec. 31.3.11
+	 */
+	return -ENOTSUP;
+}
+
+static int counter_esp32_get_value(const struct device *dev, uint32_t *ticks)
+{
+	ARG_UNUSED(dev);
+
+	SET_PERI_REG_MASK(RTC_CNTL_TIME_UPDATE_REG, RTC_CNTL_TIME_UPDATE);
+#if defined(CONFIG_SOC_ESP32)
+	while (GET_PERI_REG_MASK(RTC_CNTL_TIME_UPDATE_REG, RTC_CNTL_TIME_VALID) == 0) {
+		/* might take 1 RTC slowclk period, don't flood RTC bus */
+		k_sleep(K_USEC(1));
```

Comment:
Why not use busy_wait here?

Reply:
It may be a matter of personal taste, but the intent looks clearer to me this way.